src/AutoMapper/TypePair.cs

```diff
@@ -32,7 +30,7 @@ public override bool Equals(object obj)
             return obj is MapRequest && Equals((MapRequest) obj);
         }
 
-        public override int GetHashCode() => _hashcode;
+        public override int GetHashCode() => HashCodeCombiner.Combine(RequestedTypes, RuntimeTypes);
```

Comment:
Hmm I wouldn't say that Tuple did it guarantees that not caching it give the best performance.
Either way I don't think it's THAT critical to performance.  The only way to tell is make couple hundred type maps and map each one and then you can see if performance degrades, as it checks each has code for every type map and map request to find the execution path.

Also I don't see a reason not to cache it, because I really think if AM creates any of these mutable classes TypePair, MemberRequest, ect they WILL have to call GetHashCode() because you will be checking/adding to the _typeMapCache and every time you go through it the hashcodes will have to be called, on both sides.